### ISSUE TITLE
Development experience for contributing core functions

### ISSUE THREAD

**Issue by michael** [stencila/stencila#357]
I've been looking at the tests functionality of functions, and wondered about how the development workflow for new core functions would look like.

For instance, if I were to develop a more sophisticated plot function, in which environment would I develop it? The tests functionality for functions is more about high level testing as far as I understand it. Like:

```xml
  <tests>
    <test>
      <args>
        <arg type="number">3.14</arg>
      </args>
      <result type="string">number</result>
    </test>
```

Like providing inputs and outputs and running the function with different arguments (specified in a `<test>` element) and checking the result. 

A few concerns on this approach:

- how would I test more complicated things than simple values, e.g. in the visualization case inspect a certain json output?
- would i need to change the code and then run the Stencila test suite as a whole each time I make a change?
- Tests are interpreted, so if something fails, I guess I would not be able to see what file or what line broke the test, also could I have a debugger stop at where an error is thrown?

I thought maybe what I would prefer is developing Javascript functions in a small Javascript environment independently of the whole Stencila stack. Like just having a myplot.js and a little test setup or even an index.html loading Plot.ly to see what charts I produce. Once things are working in JS land, I would create the myplot.func.xml file and connect it to the Stencila standard library. Actually I don't like what I just proposed.. but I don't have better ideas. I'm just wondering how we can approach this so it feels good to develop functions in different languages. 

I think we can't just say create an XML file modify it.. restart Stencila and enter your function in a new Mini cell. The higher level testing is a good thing for CI but does not help with debugging during development of functions. Also the test functionality is limited, and I guess we don't want to build an advanced declarative testing framework expressed as XML?

Just thoughts.....

**Comment by nokome** [comment]
I don't think we should try to make the `<tests>` element in the `FunctionSchema` a full blown test harness that will cover all use cases. It's more intended for users who are not software developers to easily write some tests for their function within a `FunctionEditor`.

More experienced coders, writing more advanced functions would probably want to develop and test a myplot.js file using their favourite toolset and then `<code include="myplot.js" `>` in myplot.func.xml along with meta data as you suggest.

We need to code both use cases.

**Comment by nokome** [comment]
Closing this as it is largely superseded by our new approach using function libraries.
